fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LuceneLockFailureTest.java

```diff
@@ -0,0 +1,261 @@
+            openStoreWithPrefixes(context, COMPLEX_DOC, COMPLEX_PARTITIONED);
+            grabLockExternallyForPartition(COMPLEX_PARTITIONED, context, 1);
+            context.commit();
+        }
+
+        try (final FDBRecordContext context = openContext()) {
+            openStoreWithPrefixes(context, COMPLEX_DOC, COMPLEX_PARTITIONED);
+            // DeleteWhere does not check the locks, but deletes the entire group's keyspace
+            recordStore.deleteRecordsWhere(COMPLEX_DOC, Query.field("group").equalsValue(1));
+            context.commit();
+        }
+
+        try (final FDBRecordContext context = openContext()) {
+            openStoreWithPrefixes(context, COMPLEX_DOC, COMPLEX_PARTITIONED);
+            // DeleteWhere should have cleared the locks, so we should be able to take one again
+            grabLockExternallyForPartition(COMPLEX_PARTITIONED, context, 1);
+            context.commit();
+        }
+    }
+
+    @Test
+    void testMerge() throws IOException {
+        try (final FDBRecordContext context = openContext()) {
+            rebuildIndexMetaData(context, SIMPLE_DOC, SIMPLE_INDEX);
+            recordStore.saveRecord(createSimpleDocument(1623L, ENGINEER_JOKE, 2));
+            context.commit();
+        }
+
+        try (final FDBRecordContext context = openContext()) {
+            rebuildIndexMetaData(context, SIMPLE_DOC, SIMPLE_INDEX);
+            grabLockExternally(SIMPLE_INDEX, context);
+            context.commit();
+        }
+
+        try (final FDBRecordContext context = openContext()) {
+            rebuildIndexMetaData(context, SIMPLE_DOC, SIMPLE_INDEX);
+            // This fails since the merge tries to take a lock
+            Assertions.assertThrows(FDBExceptions.FDBStoreLockTakenException.class, () ->
+                    LuceneIndexTestUtils.mergeSegments(recordStore, SIMPLE_INDEX));
+        }
+    }
+
+    @Test
+    void testRebalance() throws IOException {
+        try (final FDBRecordContext context = openContext()) {
+            rebuildIndexMetaData(context, COMPLEX_DOC, COMPLEX_PARTITIONED);
+            // partition size is 10
+            for (int i = 0; i < 50; i++) {
+                recordStore.saveRecord(createComplexDocument(6666L + i, ENGINEER_JOKE, 1, Instant.now().toEpochMilli()));
+            }
+            context.commit();
+        }
+
+        try (final FDBRecordContext context = openContext()) {
+            rebuildIndexMetaData(context, COMPLEX_DOC, COMPLEX_PARTITIONED);
+            grabLockExternallyForPartition(COMPLEX_PARTITIONED, context, 1);
+            context.commit();
+        }
+
+        try (final FDBRecordContext context = openContext()) {
+            rebuildIndexMetaData(context, COMPLEX_DOC, COMPLEX_PARTITIONED);
+            // This fails since the repartition tries to take a lock
+            // The exception here is the RecordCore wrapper around the Lucene exception
+            Assertions.assertThrows(RecordCoreException.class, () ->
```

Comment:
`assertThrows` returns the exception, so you could assert that the cause is correct.

Reply:
Done. Makes sense.